deployment/7_upgrade_v2.js

```diff
@@ -0,0 +1,133 @@
+    const newVersion = {};
+
+    ////////////////////////////////////
+    // Setup
+    ////////////////////////////////////
+
+    const manager = new DeployManager(network);
+	await manager.setup();
+
+	const configurator = manager.configurator;
+	const deployer = manager.deployer;
+    const abiUploader = manager.abiUploader;
+    const versionUploader = manager.versionUploader;
+    const deploymentWallet = deployer.signer;
+    const config = configurator.config;
+
+    const ModuleRegistryWrapper = await deployer.wrapDeployedContract(ModuleRegistry, config.contracts.ModuleRegistry);
+    const MultiSigWrapper = await deployer.wrapDeployedContract(MultiSig, config.contracts.MultiSigWallet);
+    const multisigExecutor = new MultisigExecutor(MultiSigWrapper, deploymentWallet, config.multisig.autosign);
+    
+    console.log('Config:', config);
+
+    ////////////////////////////////////
+    // Deploy new modules
+    ////////////////////////////////////
+
+    const NftTransferWrapper = await deployer.deploy(
+        NftTransfer,
+        {},
+        config.contracts.ModuleRegistry,
+        config.modules.GuardianStorage
+    );
+    newModuleWrappers.push(NftTransferWrapper);  
+    
+    ///////////////////////////////////////////////////
+    // Update config and Upload new module ABIs
+    ///////////////////////////////////////////////////
+
+    configurator.updateModuleAddresses({
+        NftTransfer: NftTransferWrapper.contractAddress
+    });
+
+    const gitHash = require('child_process').execSync('git rev-parse HEAD').toString('utf8').replace(/\n$/, '');
+    configurator.updateGitHash(gitHash);
+    await configurator.save();
+
+    await Promise.all([
+        abiUploader.upload(NftTransferWrapper, "modules")
+    ]);
+
+    ////////////////////////////////////
+    // Register new modules
+    ////////////////////////////////////
+
+    for (let idx = 0; idx < newModuleWrappers.length; idx++) {
+        let wrapper = newModuleWrappers[idx];
+        await multisigExecutor.executeCall(ModuleRegistryWrapper, "registerModule", [wrapper.contractAddress, utils.asciiToBytes32(wrapper._contract.contractName)]);
+    }
+
+    ////////////////////////////////////
+    // Deploy and Register upgraders
+    ////////////////////////////////////
+
+    const toAdd = newModuleWrappers.map((wrapper) => {
+        return {
+            'name': wrapper._contract.contractName,
+            'address': wrapper.contractAddress
+        };
+    });
+    let fingerprint; 
+
+    const versions = await versionUploader.load(BACKWARD_COMPATIBILITY);
+    for (let idx = 0; idx < versions.length; idx++) {
+        const version = versions[idx];
+        const moduleNames = MODULES_TO_DISABLE.concat(MODULES_TO_ENABLE);
+        const toRemove = version.modules.filter(module => moduleNames.includes(module.name));
+        const toKeep = version.modules.filter(module => !moduleNames.includes(module.name));
+        if(idx == 0) {
```

Comment:
Why this if?

Reply:
Because we are potentially loading multiple version objects (if `BACKWARD_COMPATIBILITY` > 1). `idx == 0` means we are working on the latest version and we use that version to create the next version object that will be uploaded to S3.

For the older versions (with `idx > 0`) we only need to create and deploy the Uploader contract.

Reply:
ok